File: tests/test_volume_effects.cpp
#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/deserialization.h>
#include <opentimelineio/externalReference.h>
#include <opentimelineio/volumeEffects.h>

namespace otime = opentime::OPENTIME_VERSION;
namespace otio  = opentimelineio::OPENTIMELINEIO_VERSION;

int
main(int argc, char** argv)
{
    Tests tests;
    tests.add_test("test_audio_volume_read", [] {
        using namespace otio;

        otio::ErrorStatus              status;
        SerializableObject::Retainer<> so =
            SerializableObject::from_json_string(
                R"(
            {
                "OTIO_SCHEMA": "Clip.1",
                "media_reference": {
                    "OTIO_SCHEMA": "ExternalReference.1",
                    "target_url": "unit_test_url",
                    "available_range": {
                        "OTIO_SCHEMA": "TimeRange.1",
                        "duration": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 8
                        },
                        "start_time": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 10
                        }
                    }
                },
                "effects": [
                    {
                        "OTIO_SCHEMA": "AudioVolume.1",
                        "name": "volume",
                        "gain": 0.5,
                        "effect_name": "AudioVolume",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "AudioFade.1",
                        "name": "fade",
                        "fade_in": false,
                        "start_time": 1.5,
                        "duration": 5.0,
                        "effect_name": "AudioFade",
                        "enabled": true
                    }
                ]
            })",
                &status);

        assertFalse(is_error(status));

        const Clip* clip = dynamic_cast<const Clip*>(so.value);
        assertNotNull(clip);

        auto effects = clip->effects();
        assertEqual(effects.size(), 2);

        auto audio_volume = dynamic_cast<const AudioVolume*>(effects[0].value);
        assertNotNull(audio_volume);
        assertEqual(audio_volume->gain(), 0.5);

        auto audio_fade = dynamic_cast<const AudioFade*>(effects[1].value);
        assertNotNull(audio_fade);
        assertEqual(audio_fade->fade_in(), false);
        assertEqual(audio_fade->start_time(), 1.5);
        assertEqual(audio_fade->duration(), 5.0);
    });

    tests.add_test("test_audio_volume_write", [] {
        using namespace otio;

        SerializableObject::Retainer<otio::Clip> clip(new otio::Clip(
            "unit_clip",
            new otio::ExternalReference("unit_test_url"),
            std::nullopt,
            otio::AnyDictionary(),
            { new otio::AudioVolume("volume", 0.75),
              new otio::AudioFade("fade", true, 2.0, 10.5)}));

        auto json = clip.value->to_json_string();

        std::string expected_json = R"({
    "OTIO_SCHEMA": "Clip.2",
    "metadata": {},
    "name": "unit_clip",
    "source_range": null,
    "effects": [
        {
            "OTIO_SCHEMA": "AudioVolume.1",
            "metadata": {},
            "name": "volume",
            "effect_name": "AudioVolume",
            "enabled": true,
            "gain": 0.75
        },
        {
            "OTIO_SCHEMA": "AudioFade.1",
            "metadata": {},
            "name": "fade",
            "effect_name": "AudioFade",
            "enabled": true,
            "fade_in": true,
            "start_time": 2.0,
            "duration": 10.5
        }
    ],
    "markers": [],
    "enabled": true,
    "media_references": {
        "DEFAULT_MEDIA": {
            "OTIO_SCHEMA": "ExternalReference.1",
            "metadata": {},
            "name": "",
            "available_range": null,
            "available_image_bounds": null,
            "target_url": "unit_test_url"
        }
    },
    "active_media_reference_key": "DEFAULT_MEDIA"
})";

        assertEqual(json, expected_json);

    });

    tests.run(argc, argv);
    return 0;
}

File: tests/test_volume_effects.py
"""Volume effects class test harness."""

import unittest

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils

class AudioVolumeTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        scale = otio.schema.AudioVolume(
            name="volume",
            gain=2.5,
            metadata={
                "foo": "bar"
            }
        )
        self.assertEqual(scale.gain, 2.5)
        self.assertEqual(scale.name, "volume")
        self.assertEqual(scale.metadata, {"foo": "bar"})

    def test_eq(self):
        scale1 = otio.schema.AudioVolume(
            name="volume",
            gain=2.5,
            metadata={
                "foo": "bar"
            }
        )
        scale2 = otio.schema.AudioVolume(
            name="volume",
            gain=2.5,
            metadata={
                "foo": "bar"
            }
        )
        self.assertIsOTIOEquivalentTo(scale1, scale2)

    def test_serialize(self):
        scale = otio.schema.AudioVolume(
            name="volume",
            gain=0.6,
            metadata={
                "foo": "bar"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(scale)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(scale, decoded)

    def test_setters(self):
        scale = otio.schema.AudioVolume(
            name="volume",
            gain=0.8,
            metadata={
                "foo": "bar"
            }
        )
        self.assertEqual(scale.gain, 0.8)
        scale.gain = 0.25
        self.assertEqual(scale.gain, 0.25)

class AudioFadeTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        fade = otio.schema.AudioFade(
            name="fade",
            fade_in=True,
            start_time=12.0,
            duration=8.0,
            metadata={"baz": "qux"}
        )
        self.assertEqual(fade.name, "fade")
        self.assertEqual(fade.fade_in, True)
        self.assertEqual(fade.start_time, 12.0)
        self.assertEqual(fade.duration, 8.0)
        self.assertEqual(fade.metadata, {"baz": "qux"})

    def test_eq(self):
        fade1 = otio.schema.AudioFade(
            name="fade",
            fade_in=False,
            start_time=5.0,
            duration=3.0,
            metadata={"baz": "qux"}
        )
        fade2 = otio.schema.AudioFade(
            name="fade",
            fade_in=False,
            start_time=5.0,
            duration=3.0,
            metadata={"baz": "qux"}
        )
        self.assertIsOTIOEquivalentTo(fade1, fade2)

    def test_serialize(self):
        fade = otio.schema.AudioFade(
            name="fade",
            fade_in=True,
            start_time=2.5,
            duration=1.5,
            metadata={"baz": "qux"}
        )
        encoded = otio.adapters.otio_json.write_to_string(fade)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(fade, decoded)

    def test_setters(self):
        fade = otio.schema.AudioFade(
            name="fade",
            fade_in=False,
            start_time=4.0,
            duration=2.0,
            metadata={"baz": "qux"}
        )
        self.assertEqual(fade.fade_in, False)
        self.assertEqual(fade.start_time, 4.0)
        self.assertEqual(fade.duration, 2.0)
        fade.fade_in = True
        fade.start_time = 7.5
        fade.duration = 3.5
        self.assertEqual(fade.fade_in, True)
        self.assertEqual(fade.start_time, 7.5)
        self.assertEqual(fade.duration, 3.5)
